Stop passing -safe-string when building the compiler and in the testsuite

shindere opened the pull request:
This PR is a follow-up to a suggestion by @dra27 in
https://github.com/ocaml/ocaml/pull/11420#discussion_r918727738

PR #1252 has made the safe-string mode the default so passing -safe-string
explicitly has become useless since this PR has been merged, even more
useless since support for mutable strings was removed in OCaml 5.0.

gasche commented on manual/src/cmds/intf-c.etex:
```diff
@@ -727,8 +727,8 @@ $\hbox{"string_length"}(v)-1$.
-\var{v}, with type "char *" or, when OCaml is configured with
-"-force-safe-string", with type "const char *".
+\var{v}, with type "const char *" since strings can only be immutable from
+OCaml 5.0 onwards.
```
I would remove the historical mention, which I think is slightly confusing and not so important here.

gasche reviewed (approved):
Obviously a sensible change.

gasche commented:
(I think it's better to backport to 5.0 to also remove the `-force-safe-string` mention from the 5.0 documentation.)

shindere committed:
Stop passing -safe-string when building the compiler and in the testsuite

PR #1252 has made the safe-string mode the default so passing -safe-string
explicitly has become useless since this PR has been merged, even more
useless since support for mutable strings was removed in OCaml 5.0.

shindere commented:
Gabriel Scherer (2022/09/11 07:52 -0700):
> @gasche approved this pull request.
>
> Obviously a sensible change.

Many thanks for the prompt and positive feedback.

> > @@ -727,8 +727,8 @@ $\hbox{"string_length"}(v)-1$.
>  or byte sequence \var{v}, with type "unsigned char". Bytes are
>  numbered from 0 to $\hbox{"string_length"}(v)-1$.
>  \item "String_val("\var{v}")" returns a pointer to the first byte of the string
> -\var{v}, with type "char *" or, when OCaml is configured with
> -"-force-safe-string", with type "const char *".
> +\var{v}, with type "const char *" since strings can only be immutable from
> +OCaml 5.0 onwards.
>
> I would remove the historical mention, which I think is slightly
> confusing and not so important here.

Done.

shindere commented:
Gabriel Scherer (2022/09/11 07:53 -0700):
> (I think it's better to backport to 5.0 to also remove the
> `-force-safe-string` mention from the 5.0 documentation.)

Beware: it will likely not be just a cherry-pick, because of the changes
in the build system.

gasche commented:
Merged! I propose to push to 5.0 only the manual change.

shindere commented:
Gabriel Scherer (2022/09/11 11:28 -0700):
> Merged! I propose to push to 5.0 only the manual change.

Ah yes! Why not. Too bad I didn't think about it, I'd have odne two
separate commits, would have made things easier.

Just to make sure: did you offer to do this, or did you suggest that we
do it (no problem in the latter case, I can do it).

gasche commented:
I just did it in 4e8c5e578d37ab308a16c00ed85f8ebfcb84d39d.

Note: there remains an occurrence of `-safe-string` in the manual, in the flambda section, that one could/should consider fixing since strings really are immutable now.

https://github.com/ocaml/ocaml/blob/5cce2cc6d14cee01b0c43c9090a14d29e26bf5a7/manual/src/cmds/flambda.etex#L809-L811

However if I understand correctly the correct fix right now would be to say "flambda does not optimize... despite strings now being immutable", and the perfect fix would be to change flambda to optimize and also change the documentation. So this probably requires an flambda change, room for another PR. cc @lthls.

shindere commented:
Gabriel Scherer (2022/09/11 11:35 -0700):
> I just did it in 4e8c5e578d37ab308a16c00ed85f8ebfcb84d39d.

Thank you so much!

> Note: there remains an occurrence of `-safe-string` in the manual, in the flambda section, that one could/should consider fixing since strings really are immutable now.
>
> https://github.com/ocaml/ocaml/blob/5cce2cc6d14cee01b0c43c9090a14d29e26bf5a7/manual/src/cmds/flambda.etex#L809-L811
>
> However if I understand correctly the correct fix right now would be to say "flambda does not optimize... despite strings now being immutable", and the perfect fix would be to change flambda to optimize and also change the documentation. So this probably requires an flambda change, room for another PR. cc @lthls.

Yes, I was indeed too shy to touch it but still, my bad, apologies for
not having mentionned it explicitly in the PR.
